crates/flux-middle/src/rustc/lowering.rs

```diff
@@ -74,6 +74,45 @@ impl UnsupportedErr {
     }
 }
 
+fn resolve_call_query<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    selcx: &mut SelectionContext<'_, 'tcx>,
+    param_env: ParamEnv<'tcx>,
+    callee_id: DefId,
+    args: rustc_middle::ty::GenericArgsRef<'tcx>,
+) -> Option<(DefId, rustc_middle::ty::GenericArgsRef<'tcx>)> {
+    let mut try_resolve = || {
+        if let Some(trait_id) = tcx.trait_of_item(callee_id) {
+            let trait_ref = rustc_ty::TraitRef::from_method(tcx, trait_id, args);
+            let obligation = Obligation::new(tcx, ObligationCause::dummy(), param_env, trait_ref);
+            let impl_source = selcx.select(&obligation).ok()??;
+            let impl_source = selcx.infcx.resolve_vars_if_possible(impl_source);
+            let ImplSource::UserDefined(impl_data) = impl_source else { return None };
+
+            let assoc_id = tcx
+                .impl_item_implementor_ids(impl_data.impl_def_id)
+                .get(&callee_id)?;
+            let assoc_item = tcx.associated_item(assoc_id);
+
+            Some((assoc_item.def_id, impl_data.args))
+        } else {
+            None
+        }
+    };
+    try_resolve()
```

Comment:
Since you are inside a function that returns `Option` you can get rid of `try_resolve`

```suggestion
    if let Some(trait_id) = tcx.trait_of_item(callee_id) {
        let trait_ref = rustc_ty::TraitRef::from_method(tcx, trait_id, args);
        let obligation = Obligation::new(tcx, ObligationCause::dummy(), param_env, trait_ref);
        let impl_source = selcx.select(&obligation).ok()??;
        let impl_source = selcx.infcx.resolve_vars_if_possible(impl_source);
        let ImplSource::UserDefined(impl_data) = impl_source else { return None };

        let assoc_id = tcx
            .impl_item_implementor_ids(impl_data.impl_def_id)
            .get(&callee_id)?;
        let assoc_item = tcx.associated_item(assoc_id);

        Some((assoc_item.def_id, impl_data.args))
    } else {
        None
    }
```

Reply:
even simpler

```suggestion
    let trait_id = tcx.trait_of_item(callee_id)?;
    let trait_ref = rustc_ty::TraitRef::from_method(tcx, trait_id, args);
    let obligation = Obligation::new(tcx, ObligationCause::dummy(), param_env, trait_ref);
    let impl_source = selcx.select(&obligation).ok()??;
    let impl_source = selcx.infcx.resolve_vars_if_possible(impl_source);
    let ImplSource::UserDefined(impl_data) = impl_source else { return None };

    let assoc_id = tcx
        .impl_item_implementor_ids(impl_data.impl_def_id)
        .get(&callee_id)?;
    let assoc_item = tcx.associated_item(assoc_id);

    Some((assoc_item.def_id, impl_data.args))
```